crates/proto/src/op/message.rs

```diff
@@ -655,66 +636,56 @@ impl Message {
     pub fn read_records(
         decoder: &mut BinDecoder<'_>,
         count: usize,
         is_additional: bool,
-    ) -> ProtoResult<(Vec<Record>, Option<Edns>, Vec<Record>)> {
+    ) -> ProtoResult<(Vec<Record>, Option<Edns>, MessageSignature)> {
         let mut records: Vec<Record> = Vec::with_capacity(count);
         let mut edns: Option<Edns> = None;
-        let mut sigs: Vec<Record> = Vec::with_capacity(if is_additional { 1 } else { 0 });
+        let mut sig = MessageSignature::default();
 
-        // sig0 must be last, once this is set, disable.
-        let mut saw_sig0 = false;
-        // tsig must be last, once this is set, disable.
-        let mut saw_tsig = false;
         for _ in 0..count {
             let record = Record::read(decoder)?;
-            if saw_tsig {
-                return Err("tsig must be final resource record".into());
-            } // TSIG must be last and multiple TSIG records are not allowed
-            if !is_additional {
-                if saw_sig0 {
-                    return Err("sig0 must be final resource record".into());
-                } // SIG0 must be last
-                records.push(record)
-            } else {
-                match record.record_type() {
-                    #[cfg(feature = "__dnssec")]
-                    RecordType::SIG => {
-                        saw_sig0 = true;
-                        sigs.push(record);
-                    }
-                    #[cfg(feature = "__dnssec")]
-                    RecordType::TSIG => {
-                        if saw_sig0 {
-                            return Err("sig0 must be final resource record".into());
-                        } // SIG0 must be last
-                        saw_tsig = true;
-                        sigs.push(record);
-                    }
-                    RecordType::OPT => {
-                        if saw_sig0 {
-                            return Err("sig0 must be final resource record".into());
-                        } // SIG0 must be last
-                        if edns.is_some() {
-                            return Err("more than one edns record present".into());
-                        }
-                        edns = Some((&record).into());
-                    }
-                    _ => {
-                        if saw_sig0 {
-                            return Err("sig0 must be final resource record".into());
-                        } // SIG0 must be last
-                        records.push(record);
+
+            // There must be no additional records after a TSIG/SIG(0) record.
+            if sig != MessageSignature::Unsigned {
+                return Err("TSIG or SIG(0) record must be final resource record".into());
+            }
+
+            // SIG and TSIG records are only allowed in the additional section.
+            if !is_additional && matches!(record.record_type(), RecordType::SIG | RecordType::TSIG)
```

Comment:
I initially thought that `RecordType::OPT` should also be included in this check, but doing that broke two compatibility suite tests (`tsig_tests::test_create`, `tsig_tests::test_tsig_zone_transfer`) and since I'm both not very familiar with EDNS and the existing parser logic didn't enforce this I left it out for now.

Reply:
Aha! Thanks. I'll be mucking around in that area soon and can take a look.